TACC/Frontera-Portal
# task/FP-501-onboarding--keyservice port

## Overview: ##

Ports system creation reactor (keyservice) call functionality from CEPv1. (Does not re-implement system creation workflow, only the utility functions. This will be addressed in FP-502.

## PR Status: ##


## Related Jira tickets: ##

* [FP-701](https://jira.tacc.utexas.edu/browse/FP-701)
* [FP-501](https://jira.tacc.utexas.edu/browse/FP-501)

## Summary of Changes: ##

- task/FP-701 (base of this PR): Implements a generic webhook callback system that generates unique callback URLs, records callback data, and automatically invokes associated callbacks via the unique callback URL
- task/FP-501: Creates a utility function (call_reactor) for calling the System Creation Reactor

## Testing Steps: ##
1. Start ngrok and add the callback URL to your _WH_BASE_URL settings_secret
2. Open a python shell in the container
3. Try a dry run invoking the system creation service:

```
from django.contrib.auth import get_user_model
from portal.apps.system_creation.dryrun import dryrun_success, dryrun_failure

user = get_user_model().objects.get(username="myusername")
dryrun_success(user)
dryrun_failure(user)
```

You should see a success and a failure callback in your docker logs:

Success:
```
frontera_prtl_django | [DJANGO] INFO 2020-09-28 21:49:01,930 UTC dryrun portal.apps.system_creation.dryrun.callback:16: webhook_request: {'username': 'jchuah', 'result': 'success', 'executionId': 'JALAYNOA1EpAX', 'system': {'status': 'UP', 'description': '', 'storage': {'protocol': 'SFTP', 'homeDir': '/', 'host': 'cloud.corral.tacc.utexas.edu', 'auth': {'username': 'jchuah', 'publicKey': 'PUBLICKEY', 'privateKey': 'PRIVATEKEY', 'type': 'SSHKEYS'}, 'rootDir': 'TEST_WORK_DIR', 'port': 2222}, 'site': 'tacc.utexas.edu', 'type': 'STORAGE', 'id': 'data-tacc-work-jchuah', 'name': 'TACC $WORK Directory for jchuah'}}
```

Failure:
```
frontera_prtl_django | [DJANGO] INFO 2020-09-28 21:48:31,198 UTC dryrun portal.apps.system_creation.dryrun.callback:16: webhook_request: {'username': 'jchuah', 'result': 'Traceback (most recent call last):\n  File "/reactor.py", line 101, in createSystemDescription\n    systemDescription = agaveSystem.render(**kwargs)\n  File "/usr/local/lib/python3.6/dist-packages/jinja2/asyncsupport.py", line 76, in render\n    return original_render(self, *args, **kwargs)\n  File "/usr/local/lib/python3.6/dist-packages/jinja2/environment.py", line 1008, in render\n    return self.environment.handle_exception(exc_info, True)\n  File "/usr/local/lib/python3.6/dist-packages/jinja2/environment.py", line 780, in handle_exception\n    reraise(exc_type, exc_value, tb)\n  File "/usr/local/lib/python3.6/dist-packages/jinja2/_compat.py", line 37, in reraise\n    raise value.with_traceback(tb)\n  File "<template>", line 20, in top-level template code\njinja2.exceptions.UndefinedError: \'rootDir\' is undefined\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File "/reactor.py", line 241, in main\n    agaveSystem, systemID, sanitized = createSystemDescription(templateFile, **variables)\n  File "/reactor.py", line 104, in createSystemDescription\n    raise Exception("Template error, {}".format(str(e)))\nException: Template error, \'rootDir\' is undefined\n', 'executionId': 'QbeRpL00ppO16', 'system': None}
```

## UI Photos:

## Notes: ##
